tls/s2n_signature_algorithms.c

```diff
@@ -124,6 +140,64 @@ static int s2n_tls13_default_sig_scheme(struct s2n_connection *conn, struct s2n_
     S2N_ERROR(S2N_ERR_INVALID_SIGNATURE_SCHEME);
 }
 
+static int s2n_default_sig_scheme_server(struct s2n_connection *conn, struct s2n_signature_scheme *sig_scheme_out)
+{
+    notnull_check(conn);
+    notnull_check(conn->secure.cipher_suite);
+    notnull_check(sig_scheme_out);
+
+    s2n_authentication_method cipher_suite_auth_method = conn->secure.cipher_suite->auth_method;
+    S2N_ERROR_IF(
+        (cipher_suite_auth_method != S2N_AUTHENTICATION_ECDSA && cipher_suite_auth_method != S2N_AUTHENTICATION_RSA),
+        S2N_ERR_INVALID_SIGNATURE_SCHEME);
+
+    /* Default our signature digest algorithms. For TLS 1.2 this default is different and may be
+     * overridden by the signature_algorithms extension. If the server chooses an ECDHE_ECDSA
+     * cipher suite, this will be overridden to SHA1.
+     */
+    *sig_scheme_out = s2n_rsa_pkcs1_md5_sha1;
+
+    if (cipher_suite_auth_method == S2N_AUTHENTICATION_ECDSA) {
+        *sig_scheme_out = s2n_ecdsa_sha1;
+    }
+
+    /* Default RSA Hash Algorithm is SHA1 (instead of MD5_SHA1) if TLS 1.2 or FIPS mode */
+    if ((conn->actual_protocol_version >= S2N_TLS12 || s2n_is_in_fips_mode())
+            && (sig_scheme_out->sig_alg == S2N_SIGNATURE_RSA)) {
+        *sig_scheme_out = s2n_rsa_pkcs1_sha1;
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_default_sig_scheme_client(struct s2n_connection *conn, struct s2n_signature_scheme *sig_scheme_out)
+{
+    notnull_check(conn);
+    notnull_check(sig_scheme_out);
+
+    s2n_pkey_type client_cert_pkey_type = conn->secure.client_cert_pkey_type;
```

Comment:
Is where the s2n_pkey_type comes from the only difference between these two methods? If so, why not just make it an argument to the s2n_default_sig_scheme method instead of having separate methods for client and server?

Reply:
Yes, the difference between the client and server is that we use the `client_cert_pkey_type` instead of `cipher_suite_auth_method` to determine the default signature scheme.  These are static functions that will be called within `s2n_choose_default_sig_scheme` function based on the mode, the reason for separating them to different methods was for readability and separation of logic into logical blocks instead of using multiple if-else statements within a single function.

Reply:
You lose readability to the need to compare the two methods for minor differences. You've got duplicate code, and any changes to the default logic will need to be made to both methods. Instead of choosing which method to call, let the caller choose which argument to pass in.